contract-v3/src/main.rs

```diff
@@ -0,0 +1,237 @@
+
+    // Store the contract version in the context's named keys.
+    let version_uref = storage::new_uref(contract_version);
+    runtime::put_key(CONTRACT_VERSION_KEY, version_uref.into());
+
+    // Create a named key for the contract hash.
+    runtime::put_key(CONTRACT_KEY, stored_contract_hash.into());
+}
+
+/// Helper function that upgrades the contract package to a new version.
+fn upgrade_counter() {
+    let mut counter_entry_points = EntryPoints::new();
+    counter_entry_points.add_entry_point(EntryPoint::new(
+        ENTRY_POINT_COUNTER_GET,
+        Vec::new(),
+        CLType::I32,
+        EntryPointAccess::Public,
+        EntryPointType::Contract,
+    ));
+
+    counter_entry_points.add_entry_point(EntryPoint::new(
+        ENTRY_POINT_COUNTER_INC,
+        Vec::new(),
+        CLType::Unit,
+        EntryPointAccess::Public,
+        EntryPointType::Contract,
+    ));
+
+    // Create an entry point to decrement the counter by 1.
+    counter_entry_points.add_entry_point(EntryPoint::new(
+        ENTRY_POINT_COUNTER_DECREMENT,
+        Vec::new(),
+        CLType::Unit,
+        EntryPointAccess::Public,
+        EntryPointType::Contract,
+    ));
+
+    counter_entry_points.add_entry_point(EntryPoint::new(
+        ENTRY_POINT_COUNTER_LAST_UPDATED_AT,
+        Vec::new(),
+        CLType::U64,
+        EntryPointAccess::Public,
+        EntryPointType::Contract,
+    ));
+
+    // Get the counter package hash so we can upgrade the package.
+    let counter_package_hash = runtime::get_key(CONTRACT_PACKAGE_NAME)
+        .unwrap_or_revert()
+        .into_hash()
+        .unwrap()
+        .into();
+
+    // Add a new named_key
+    let mut named_keys = NamedKeys::new();
+    let last_updated = storage::new_uref(0_u64);
+    named_keys.insert(String::from(LAST_UPDATED_KEY), last_updated.into());
+
+    // Add a new contract version to the package with the new list of entry points.
+    let (stored_contract_hash, contract_version) =
+        storage::add_contract_version(counter_package_hash, counter_entry_points, named_keys);
+
+    // Here we are updating the version named key with a new value.
+    // The version named key should already be part of the account.
+    let version_uref = storage::new_uref(contract_version);
+    runtime::put_key(CONTRACT_VERSION_KEY, version_uref.into());
+
+    // Add the latest contract hash into the named key.
+    // The key should already exist and we will have access to it in this version.
+    runtime::put_key(CONTRACT_KEY, stored_contract_hash.into());
+}
+
+/// Entry point that executes automatically when a caller installs the contract.
+#[no_mangle]
+pub extern "C" fn call() {
+    match runtime::get_key(CONTRACT_ACCESS_UREF) {
+        None => {
+            // The given key doesn't exist, so install the contract.
+            install_counter();
+            // Next, upgrade the contract.
+            upgrade_counter();
```

Comment:
Why do you upgrade the contract you just deployed?

Reply:
This is done the same way as the existing `contract-v2`. I believe this is done so the code from `upgrade_contract` is not repeated in install contract function